Support no_std

wangrunji0408 opened the pull request:
Now feature `alloc` is available in stable Rust.
It's time to make it `no_std` compatible.

Similar to PR #5 .

marcianx commented:
Thank you for contributing!
I'm seeing a bunch of 

    error[E0433]: failed to resolve: maybe a missing crate `core`?

errors when running tests via `cargo test`. Could you possibly resolve these?

marcianx commented:
And just to emphasize, I'd love for this to be fixed without updating the tests themselves to test backward-compatibility.

wangrunji0408 commented:
I realized that my previous implementation was not correct. It will break on std.
So I rewrote it by referring to [this guide](https://users.rust-lang.org/t/best-practice-to-write-a-macro-for-both-std-and-no-std/13725/2).
And now it works well for both `std` and `no_std` with all tests unchanged.

I also added GitHub Actions to make sure everything is fine. You can check the result [here](https://github.com/rcore-os/downcast-rs/runs/815238791?check_suite_focus=true).

marcianx commented:
Thank, you this is excellent!

marcianx commented:
Huh, rebase didn't rebase the common commit properly. I'll fix it offline and force a push. You might have trouble syncing with your branch due to merge conflicts.
